### ISSUE TITLE
loss nan problem

### ISSUE THREAD

**Issue by lc82111** [xingjian-f/DeepLearning-OCR#1]
Hi, I got nan problem during the training. I generated captcha by making use your code. The captcha consists of 4-6 chars. Either CNN_LSTM or CNN_Merge leads to same nan problem. I also try different optimer like SGD or lower learning rate, but the problem still exists.  I past the outputs in below. Any idea?

Loading data...
Data loaded, spend time(m) : 0.0166809479396
Loading data...
Data loaded, spend time(m) : 0.417937397957
X_train shape: (50000, 1, 48, 200)
50000 train samples
Train on 50000 samples, validate on 2000 samples
Epoch 1/500
50000/50000 [==============================] - 549s - loss: nan - acc: 0.0135 - val_loss: nan - val_acc: 0.0146
Epoch 2/500
50000/50000 [==============================] - 543s - loss: nan - acc: 0.0135 - val_loss: nan - val_acc: 0.0146
Epoch 3/500
50000/50000 [==============================] - 546s - loss: nan - acc: 0.0135 - val_loss: nan - val_acc: 0.0146
Epoch 4/500
50000/50000 [==============================] - 562s - loss: nan - acc: 0.0135 - val_loss: nan - val_acc: 0.0146
Epoch 5/500
50000/50000 [==============================] - 550s - loss: nan - acc: 0.0135 - val_loss: nan - val_acc: 0.0146
Epoch 6/500
38080/50000 [=====================>........] - ETA: 129s - loss: nan - acc: 0.0136

**Comment by lc82111** [comment]
It seems the BatchNormalization layer causes the NaN problem that disappears if this layer is removed.

I will try tensorflow backend instead of theano.

**Comment by xingjian-f** [comment]
@lc82111 Could you tell me the version of keras you are using, and email me a small data set you used to train. I will try to train on it in my environment to see whether this problem will appear again. BTW, I'm using keras 1.0.6, it's not the latest version, maybe this is the issue.  And please make sure you are using the latest version of my code, thanks!

**Comment by lc82111** [comment]
@xingjian-f 
After switching to tensorflow backend, the NAN problem is gone. 
Now, I train your vgg-merge model on a dataset with 100k samples and hope everything goes well.

BTW, how many samples do you use under the vgg-merge/CNN-LSTM models to solve the fixed/varying length captcha tasks?  What sequence accuracy do you achieve?

It seems the vgg-like model with complex architecture makes the training become difficult. Is there any possible to use a simpler cnn architecture to improve the accuracy?

**Comment by xingjian-f** [comment]
@lc82111 I usually use 10~50k for both vgg-merge/cnn-lstm models and fixed/varying length tasks is also the same. It appears that varying length tasks are more difficult, so sometimes I need to gather more data for this kind of task.

Both models can achieve more than 90% sequence accuracy easily for sequence length less than 10.

Yeah, It seems the vgg-like model takes longer time to train, to use a simpler architecture, you may check this paper "Multi-digit Number Recognition from Street View Imagery using Deep Convolutional Neural Networks", a paper published by Google's reCAPTCHA Teams.

**Comment by lc82111** [comment]
@xingjian-f Thanks. I have checked the Google's paper. But, apparently, they used the generative model (deep belief net?). I find a thesis that use a shallow  CNN you may check: Neural Network Captcha Cracker
